File: exercises/subtree_of_tree_572.java
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}
public class subtree_of_tree_572 {
    public static void main(String[] args) {
        var sol = new Solution_56();
        var tree1 = new TreeNode(3, new TreeNode(4, new TreeNode(1), new TreeNode(2, new TreeNode(0), null)), new TreeNode(5));
        var tree2 = new TreeNode(4, new TreeNode(1), new TreeNode(2));
        var res = sol.isSubtree(tree1, tree2);
        System.out.printf("res: %b\n", res);
    }
}
class Solution_56 {
    public boolean isSubtree(TreeNode a, TreeNode b) {
        if (a == null && b == null) return true;
        if (a == null || b == null) return false;

        if (a.val == b.val) {
            return isSubtree(a.left, b.left) && isSubtree(a.right, b.right);
        }
        return isSubtree(a.left, b) || isSubtree(a.right, b);
    }
}

/*
 * Given an original tree and a possible subtree figure out if the second tree is a 
 * proper subtree of the original tree
 * For the second tree to be a subtree of the original tree there must be a node in 
 * the original tree which if taken as the root would be equivalent to the second tree
 */

File: exercises/subtree_of_tree_572.py
from typing import Optional
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:   
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if not root:
            return not subRoot
        return self.isEqual(root, subRoot) or self.isSubtree(root.left, subRoot) \
               or self.isSubtree(root.right, subRoot) 

    def isEqual (self, a: Optional[TreeNode], b: Optional[TreeNode]) -> bool:
        if not a:
            return not b
        if not b:
            return False
        return a.val == b.val and self.isEqual(a.left, b.left) and self.isEqual(a.right, b.right)

# Alternative solution
# class solution:   
#     def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
#         return self.subtree(root, subRoot, subRoot)
#
#     def subtree (self, root: Optional[TreeNode], subRoot: Optional[TreeNode], currSub: Optional[TreeNode]) -> bool:
#         if not root:
#             return not currSub
#         if not currSub:
#             return False
#         if root.val == currSub.val and self.subtree(root.left, subRoot, currSub.left) \
#                             and self.subtree(root.right, subRoot, currSub):
#             return True
#         if root.val == subRoot and self.subtree(root.left, subRoot, subRoot.left) \
#                             and self.subtree(root.right, subRoot, subRoot.right):
#             return True
#         return self.subtree(root.left, subRoot, subRoot) or self.subtree(root.right, subRoot, subRoot)

